airflow/io/__init__.py

```diff
@@ -0,0 +1,270 @@
+as check if a file exists. An implementation of the FileIO abstract base class is responsible
+for returning an InputFile instance, an OutputFile instance, and deleting a file given
+its location.
+
+Ported from Apache Iceberg.
+"""
+from __future__ import annotations
+
+import importlib
+import logging
+from abc import ABC, abstractmethod
+from io import SEEK_SET
+from types import TracebackType
+from typing import (
+    Dict,
+    Optional,
+    Protocol,
+    runtime_checkable,
+)
+
+log = logging.getLogger(__name__)
+
+HDFS_HOST = "hdfs.host"
+HDFS_PORT = "hdfs.port"
+HDFS_USER = "hdfs.user"
+HDFS_KERB_TICKET = "hdfs.kerberos_ticket"
+
+TOKEN = "token"
+
+Properties = Dict[str, Optional[str]]
+
+
+@runtime_checkable
+class InputStream(Protocol):
+    """A protocol for the file-like object returned by InputFile.open(...).
+
+    This outlines the minimally required methods for a seekable input stream returned from an InputFile
+    implementation's `open(...)` method. These methods are a subset of IOBase/RawIOBase.
+    """
+
+    @abstractmethod
+    def read(self, size: int = 0) -> bytes:
+        ...
+
+    @abstractmethod
+    def seek(self, offset: int, whence: int = SEEK_SET) -> int:
+        ...
+
+    @abstractmethod
+    def tell(self) -> int:
+        ...
+
+    @abstractmethod
+    def close(self) -> None:
+        ...
+
+    def __enter__(self) -> InputStream:
+        """Provide setup when opening an InputStream using a 'with' statement."""
+
+    @abstractmethod
+    def __exit__(
+        self,
+        exctype: type[BaseException] | None,
+        excinst: BaseException | None,
+        exctb: TracebackType | None,
+    ) -> None:
+        """Perform cleanup when exiting the scope of a 'with' statement."""
+
+
+@runtime_checkable
+class OutputStream(Protocol):  # pragma: no cover
+    """A protocol for the file-like object returned by OutputFile.create(...).
+
+    This outlines the minimally required methods for a writable output stream returned from an OutputFile
+    implementation's `create(...)` method. These methods are a subset of IOBase/RawIOBase.
+    """
+
+    @abstractmethod
+    def write(self, b: bytes) -> int:
+        ...
+
+    @abstractmethod
+    def close(self) -> None:
+        ...
+
+    @abstractmethod
+    def __enter__(self) -> OutputStream:
+        """Provide setup when opening an OutputStream using a 'with' statement."""
+
+    @abstractmethod
+    def __exit__(
+        self,
+        exctype: type[BaseException] | None,
+        excinst: BaseException | None,
+        exctb: TracebackType | None,
+    ) -> None:
+        """Perform cleanup when exiting the scope of a 'with' statement."""
+
+
+class InputFile(ABC):
+    """A base class for InputFile implementations.
+
+    Args:
+        location (str): A URI or a path to a local file.
+
+    Attributes:
+        location (str): The URI or path to a local file for an InputFile instance.
+        exists (bool): Whether the file exists or not.
```

Comment:
How does this render? I don’t think our Sphinx setup supports this docstring style.

Reply:
I'll fix that